[Guo Yuhao] ip

RingoftheKing opened the pull request:
# BadApple v0.1

> "Reimu has come to White Space to remind you not to Dilly Dally today" 

BadApple (or BadPingGuo) is the premiere application for a fuss free Todo list. It is:
- text based
- <s>fast</s> SPEEDY
- Real **easy** to pick up

To use it:
1. Download it [here](https://github.com/ringoftheking/ip)
2. Click on the Green Code button. Download the zip.
3. Open the zip and enter the main function
4. Run it.
5. Type commands such as todo, deadline, event.
6. It'll help you do the rest 💀 

Features currently available:
- [x] todo items. A simple todo <desc> item
- [x] Deadlines. A todo with a DATE as a deadline
- [x] Events. An item with a **start** and **end** time

To help improve our chatbot:
- [ ] Make a GUI with [JavaFx](https://se-education.org/guides/tutorials/javaFx.html)
- [ ] Help us check our coding standards
- [ ] Make Commands more robust with a [Command Pattern](https://www.tutorialspoint.com/design_pattern/command_pattern.htm). Currently a minor tweak to file structure destroys code integrity.


For those Java Savvy, You may practice `Java`  with this. For those with even more ideas you may consider helping us Abstract better classes. Here is a sample of (a relatively crowded) main function in our codebase

```java
public static void main(String[] args) {
        Ui.showWelcome();
        Scanner sc = new Scanner(System.in);
        String filename = "src/main/data/whiteSpace.txt";

        try {
            File file = new File(filename);
            FileReader fc = new FileReader(file);
            BufferedReader reader = new BufferedReader(fc);

            Tracker.suppressMessages = true;
            Storage.parseTasks(file);
            Tracker.suppressMessages = false;

            TaskList.listTasks(reader);
            System.out.println("Waiting for something to happen?");

            // self note: update the file everytime an operation is complete.
            // Delete the old file, for loop all the tasks into new file. Rename it to old file name.
            while(true) {
                String request = sc.nextLine();
                if (request.equalsIgnoreCase("bye")) break;
                Parser.ProcessQuery(request, file);
            }

            System.out.println("--------------------------------");
            System.out.println("Everything is going to be okay.");

            fc.close();
            reader.close();

        } catch (FileNotFoundException e) {
            System.out.println("make necessary files? yes/no" );
            String askToCreateFile = sc.nextLine();
            if (askToCreateFile.equalsIgnoreCase("yes")) {
                makeFile();
            }
        } catch (IOException e) {
            System.out.println("unable to process file");
        } catch (BadAppleException be) {
            System.out.println(be.toString());
        }

    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

RingoftheKing committed:
Initialise entry and exit lines

RingoftheKing committed:
Added Echoing function

RingoftheKing committed:
Added 'add' and 'List' functions

RingoftheKing committed:
added functionality to Entry Point of program

RingoftheKing committed:
Added Marking functionality, including Exception Handling

RingoftheKing committed:
added types of Tasks: TODO, Event, Deadline

RingoftheKing committed:
corrected empty files, added Task display when marking

RingoftheKing committed:
Add args handling to Tasks and inheritors

Deadline, Event, Todo accept all args after.
Incorrect args generate an empty Task and empty fields.

It should handle incorrect formatting of args

Add args handling to Task inheritors, add Exception handling in Tracker

This prevents users from entering incorrect inputs and creating
redundant entries in list. Exceptions NOT handled in tasks, future
features may need to know such exceptions. Exception handled in Tracker.

RingoftheKing committed:
Add Delete Function

RingoftheKing committed:
Configure runtest.bat to correct input file

RingoftheKing committed:
Merge branch 'master' of https://github.com/RingoftheKing/ip

RingoftheKing committed:
Relegate Task reading and writing to files

Tasks are now written to a file on the computer
(or generated if not exist)

Users can fetch data after inputing their Tasks
rather than having it disappear

txt files are easy to write to and easy to read from
making them a good way enforce permanence in data.

RingoftheKing committed:
Merge branch-level-7 with master

RingoftheKing committed:
Merge branch 'master' of https://github.com/RingoftheKing/ip

RingoftheKing committed:
Specify Dates to be used in Deadline

Deadline originally allowed any String to be set
as its /by value.

It is better to use an actual date so future filters
can search for deadlines and prioritise them

RingoftheKing committed:
Merge branch 'branch-Level-8'

RingoftheKing committed:
Add Ui, TaskList, Storage, Parser classes

Code logic was all placed in Tracker.java
This impedes new developers from understanding
what the Tracker's function is.

It is better to have a segregation of concerns
so that each feature is placed in its own classes

As such 4 new classes are made to accomodate
such functions

RingoftheKing committed:
Merge branch 'branch-more-OOP'

RingoftheKing committed:
Move classes to packages Main, Tasks

Besides relegating code logic to more classes
It is also good to move classes to packages
that clearly denote their functions

Additonal segregation of concerns thus added

RingoftheKing committed:
Merge branch 'add-gradle-support' of https://github.com/RingoftheKing/ip

RingoftheKing committed:
Add gradle build

Created a better environment for junit testing

RingoftheKing committed:
Add Testcases for Todo and Parser

It is good to do testing on methods to
ensure that future code edits do not
cause dependency failures.

Therefore we added some preliminary
test cases.

RingoftheKing committed:
Add JavaDoc to all classes

For easier understanding of what each class does.

RingoftheKing committed:
Follow SE-Edu coding standard.

Coding standards are good practices
Following them provides readability.

RingoftheKing committed:
Add Find functionality

Users can now find tasks that
match a given search string.
Matches need not be exact, for example
CS will find CS2103

RingoftheKing committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

RingoftheKing committed:
Merge remote-tracking branch 'origin/branch-Level-9'

gerteck commented on src/main/java/BadApple/main/BadPingGuo.java:
```diff
@@ -0,0 +1,80 @@
+package BadApple.main;
```
Would it be better to represent package names in all lowercase?  👿🍎

gerteck commented on src/main/java/BadApple/main/BadPingGuo.java:
```diff
@@ -0,0 +1,80 @@
+
+            while(true) {
+                String request = sc.nextLine();
+                if (request.equalsIgnoreCase("bye")) break;
```
Would it be good to wrap with curly brackets for this single statment conditional? ➰

gerteck commented on src/main/java/BadApple/main/Ui.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Ui {
+    public static void showWelcome() {
+        String logo = " __      __.__    .__  __             _________                          \n" +
```
Assuming the `logo` is a constant name, would it enhance readability by adopting uppercase with underscore notation, e.g. `LOGO`?

gerteck commented on src/main/java/BadApple/task/Messenger.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Random;
+
+public class Messenger {
+    public static boolean suppressMessages = false;
```
Could using `isMessagesSuppressed` help increase readability?

DhiraPT replied:
I agree with @gerteck. `isMessagesSuppressed` sounds more like boolean.

gerteck commented on src/main/java/BadApple/task/Storage.java:
```diff
@@ -0,0 +1,161 @@
+     * @param i the index at which a task should be removed
+     */
+    public static void removeTask(int i) {
+        System.out.println("Exploooosion! now task " + TaskList.tasks.remove(i).brief() + " has been Kazuma-ed out of existence");
```
I like how Megumin helped to remove the task! 👍

gerteck reviewed (commented):
LGTM! Added some suggestions regarding coding standards that could help enhance readability. I really like the creativity of `BadApple`!

DhiraPT commented on src/main/java/BadApple/task/Storage.java:
```diff
@@ -0,0 +1,161 @@
+
+import BadApple.main.BadAppleException;
+
+import java.io.*;
```
I don't think wildcard import is recommended. Perhaps list the imported classes explicitly?

DhiraPT commented on src/main/java/BadApple/task/Parser.java:
```diff
@@ -0,0 +1,108 @@
+        ArrayList<String> args = new ArrayList<String>(Arrays.asList(tokens));
+
+        switch (args.get(0).toLowerCase()) {
+            case "list":
```
I don't think there should be an indentation here. Perhaps you can remove the indentation for the `case` clauses? I noticed the same issue in the `Storage` class too.

DhiraPT commented on src/main/java/BadApple/task/Storage.java:
```diff
@@ -0,0 +1,161 @@
+     * @throws BadAppleException when the file contents are in the wrong format (i.e. non-command)
+     */
+    public static void parseTasks(File file) throws IOException, BadAppleException {
+        // check the file to see what tasks are already available.
```
Would it be better to begin the comment with a capital letter and use proper sentence punctuation for improved readability? I noticed the same issue for some other comments too.

DhiraPT commented on src/main/java/BadApple/task/Parser.java:
```diff
@@ -0,0 +1,108 @@
+     * @param s the query the user would like to make
+     * @param file the file to read or write from
+     */
+    public static void ProcessQuery(String s, File file) throws IOException {
```
Any reason for not using camelCase for `ProcessQuery(...)` method? Would it be better to change this to `processQuery(...)`?

RingoftheKing committed:
Add basic window to application

This preliminary version features
a simple GUI that has only a send
button and input text with fixed
frame width.

RingoftheKing committed:
Add DialogBox to application

The chatbot was unable to show
interaction with the user.

We added two DialogBox types
to differentiate the Com and
the user.

RingoftheKing committed:
Add FXML to chatbot

Windows needed to be configured manually. This makes changes tedious.

Use FXML to easily edit window
and element sizes.

RingoftheKing committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/BadApple/main/BadPingGuo.java
#	src/main/java/BadApple/task/Parser.java
#	src/main/java/BadApple/task/Storage.java
#	src/main/java/BadApple/task/Task.java
#	src/main/java/BadApple/uiElements/Ui.java

RingoftheKing committed:
Fix Missing dependencies.

Tracker and TaskList were missing key
functional features.

Reinstate and add back dependencies
for app to work again.

RingoftheKing committed:
Include Task Count Assertion

A simple assertion that counts if
the TaskList has grown by 1 when an
additive command is ran.

Leave branch to improve code quality
prior to adding more assertions.
Parser methods are too long

RingoftheKing committed:
Merge pull request #2 from RingoftheKing/A-Assertions

Include Task Count Assertion

Count Tasks when added.

RingoftheKing committed:
Shorten Parser methods

Many methods were too long especially
when reading user input

Add more classes to separate concerns
and improve readability.

RingoftheKing committed:
Shorten Methods and remove deep nests

Previous code included very long methods
especially Parsing methods.

Abstract long methods into smaller classes
this also avoids deep nesting.

RingoftheKing committed:
Merge pull request from branch-A-CodeQuality

Improve Code Quality

RingoftheKing committed:
Add assertions to TaskList and BadApple

We should test important points of code
to see if functionality is as intended.

Add assertions about TaskList changing
its size on Adds and Deletes
and assertions about File making.

RingoftheKing committed:
Merge pull request from A-Assertions

Add assertions to TaskList and BadApple

RingoftheKing committed:
Fix indexing issues with TaskList

The app will stop responding if an
Index out of bounds was used.

Implement bound checking and Ui
message to warn user of wrong index

RingoftheKing committed:
Add Update functionality

Once a user adds a task, it was
difficult to edit without deletion.

Add Update command, such that
task descriptions, due by, from
and to can be changed.

RingoftheKing committed:
Merge pull request from branch-C-Update

Add Update functionality

RingoftheKing committed:
Change README and adjust JavaFX size

Update ReadMe to contain features
of the application.

Adjust JavaFX size to fit longer text

RingoftheKing committed:
Fix Deadline parsing

It was not possible to parse an
update command without a desc

Fix Deadline.java such that a desc
is optional

RingoftheKing committed:
Fix File paths

There was an issue creating
directory and whiteSpace file

Try change to relative path

RingoftheKing committed:
Add JavaDoc to newer methods

New functions and classes were
introduced.

Included the new documentation
as such

RingoftheKing committed:
Add JavaDoc to newer methods

New functions and classes were
introduced.

Included the new documentation
as such
